python_modules/dagster/dagster/_core/definitions/asset_layer.py

```diff
@@ -906,21 +877,23 @@ def build_asset_selection_job(
                 f"{partitions_def}.",
             )
 
-    all_included_assets = [*included_assets, *included_source_assets]
-    executable_assets = [asset for asset in all_included_assets if asset.is_executable]
-    loadable_assets = [
-        *(asset for asset in all_included_assets if not asset.is_executable),
-        *(asset for asset in source_assets if asset not in included_source_assets),
-        *excluded_assets,
+    executable_assets_defs = [asset for asset in included_assets if asset.is_executable]
+    unexecutable_assets_defs = [
+        unexecutable_ad
+        for ad in (
+            *(asset for asset in included_assets if not asset.is_executable),
+            *excluded_assets,
+        )
+        for unexecutable_ad in create_unexecutable_external_assets_from_assets_def(ad)
```

Comment:
How would you feel about encoding this in the type system with a new type??

```ExecutableAssetDefs = NewType(List[AssetsDefinition], "ExecutableAssetDefs")```

I think this pattern would be nice for being more disciplined about lists of assets defs separated by "kind"

Reply:
I'd prefer not to do this because there is no enforcement that the `AssetsDefinitions` in the list would actually be executable, which could be pretty confusing.

Reply:
we would have a conversion function that checks that invariant and then returns the new type.

Reply:
Something like:

```python
from typing import Iterable, List, NewType

from dagster import AssetsDefinition

UnexecutableAssetsDefinitions = NewType("UnexecutableAssetsDefinitions", List[AssetsDefinition])


def to_unexecutable_assets_definitions(
    assets_definitions: Iterable[AssetsDefinition],
) -> UnexecutableAssetsDefinitions:
    return UnexecutableAssetsDefinitions([ad for ad in assets_definitions if not ad.is_executable])
```

Reply:
Interesting. IMO if we did this I think we'd want to use the singular, so you'd have `to_unexecutable_assets_definition` and `List[UnexecutableAssetsDefinition]`. I think we should assess when the smoke clears from this PR, the `AssetChecksDefinition` removal, and the hollowing out of `AssetLayer`.

Reply:
And yes no need to do it here.